pallets/pallet-did-lookup/src/migrations.rs

```diff
@@ -16,74 +16,151 @@
+
+	if let Some(value) = unhashed::get_raw(&old_prefix) {
+		unhashed::put_raw(&new_prefix, &value);
+		unhashed::kill(&old_prefix);
 	}
+}
 
+pub struct EthereumMigration<T>(PhantomData<T>);
+
+impl<T: crate::pallet::Config> OnRuntimeUpgrade for EthereumMigration<T>
+where
+	T::AccountId: Into<LinkableAccountId>,
+{
 	fn on_runtime_upgrade() -> frame_support::weights::Weight {
-		// Account for the new storage version written below.
-		let initial_weight = T::DbWeight::get().writes(1);
-
-		// Origin was disabled, so there cannot be any existing links. But we check just
-		// to be sure.
-		let total_weight: Weight =
-			ConnectedDids::<T>::iter().fold(initial_weight, |total_weight, (account, record)| {
-				ConnectedAccounts::<T>::insert(record.did, account, ());
-				// One read for the `ConnectedDids` entry, one write for the new
-				// `ConnectedAccounts` entry.
-				total_weight.saturating_add(T::DbWeight::get().reads_writes(1, 1))
+		if Pallet::<T>::current_storage_version() == StorageVersion::new(3) {
+			// already on version 3
+			<T as frame_system::Config>::DbWeight::get().reads_writes(1, 0)
+		} else {
+			log::info!("🔎 DidLookup: Initiating migration");
+			let mut connected_dids = 0u64;
+			let mut connected_accounts = 0u64;
+
+			// Migrate connected DIDs
+			// We should not write to the same storage item during drain because it can lead
+			// to undefined results. Thus, we write to a temporary storage and move that at
+			// the end. Else we iterate over every key more or less twice.
+			ConnectedDids::<T>::drain().for_each(|(acc_id32, value)| {
+				log::debug!(
+					"🔎 #{:?} Migrating ConnectedDid for account id {:?}",
+					connected_dids.encode(),
+					acc_id32.encode()
+				);
+				let acc_id: LinkableAccountId = acc_id32.into();
+				TmpConnectedDids::<T>::insert(acc_id, value);
+				connected_dids += 1;
+			});
+			log::info!("🔎 DidLookup: Migrated all {:?} ConnectedDids", connected_dids);
+
+			// Migrate accounts
+			ConnectedAccounts::<T>::drain().for_each(|(did_id, acc_id32, val)| {
+				log::debug!(
+					"🔎 #{:?} Migrating ConnectedAccount for did_id {:?} and account id {:?}",
+					connected_accounts,
+					did_id.encode(),
+					acc_id32.encode()
+				);
+				let acc_id: LinkableAccountId = acc_id32.into();
+				TmpConnectedAccounts::<T>::insert(did_id, acc_id, val);
+				connected_accounts += 1;
 			});
+			log::info!("🔎 DidLookup: Migrated all {:?} ConnectedAccounts", connected_accounts);
+
+			// Move TmpStorage
+			move_storage::<Pallet<T>>(b"TmpConnectedDids", b"ConnectedDids");
+			move_storage::<Pallet<T>>(b"TmpConnectedAccounts", b"ConnectedAccounts");
 
-		Pallet::<T>::current_storage_version().put::<Pallet<T>>();
+			StorageVersion::new(3).put::<Pallet<T>>();
 
+			<T as frame_system::Config>::DbWeight::get().reads_writes(
+				(connected_dids.saturating_add(connected_accounts)).saturating_mul(2),
+				(connected_dids.saturating_add(connected_accounts))
+					.saturating_mul(2)
```

Comment:
AFAIK delete is also a storage write, so this would be a 4?

Reply:
clarified in https://github.com/KILTprotocol/kilt-node/pull/355/commits/f049707cb4bd3d7486a6e8424eca79009d2908b1

Reply:
That's way better with comments. This way I noticed that we missed the `+1` for reading the storage version? fcca140a40a6c40e99f45931fe2a99636b97a4f8